clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -258,6 +258,11 @@ private static FetchResponseData toMessage(Errors error,
             FetchResponseData.PartitionData partitionData = entry.getValue();
             // Since PartitionData alone doesn't know the partition ID, we set it here
             partitionData.setPartitionIndex(entry.getKey().topicPartition().partition());
+            // To protect the clients from failing due to null records,
```

Comment:
I was thinking of the following. Make the following constructor private and change FetchRequest.getErrorResponse() to use the `of` constructor.

`    public FetchResponse(FetchResponseData fetchResponseData)`

Add a comment that the `of` constructor is for the server and the `parse` constructor is for the client.

@chia7712 : Is the following TODO still valid?

```
    // TODO: remove as a part of KAFKA-12410
    public static FetchResponse of(Errors error,

```

Reply:
>  Is the following TODO still valid?

yes, it is valid. the ticket is used to avoid looping over all data when building `FetchResponseData`. I have opened https://issues.apache.org/jira/browse/KAFKA-19053 to remove one of `FetchResponse#of`, but another `FetchResponse#of` is still used by `FetchSession`